bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Control.java

```diff
@@ -4900,28 +4900,32 @@ LRESULT WM_DESTROY (long wParam, long lParam) {
 LRESULT WM_DPICHANGED (long wParam, long lParam) {
 	// Map DPI to Zoom and compare
 	int newNativeZoom = DPIUtil.mapDPIToZoom (OS.HIWORD (wParam));
-	int oldNativeZoom = getShell().nativeZoom;
-
-	// Throw the DPI change event if zoom value changes
-	if (newNativeZoom != oldNativeZoom) {
-		Event event = new Event();
-		event.type = SWT.ZoomChanged;
-		event.widget = this;
-		event.detail = newNativeZoom;
-		event.doit = true;
-
-		if (getDisplay().isRescalingAtRuntime()) {
+	if (getDisplay().isRescalingAtRuntime()) {
+		int oldNativeZoom = nativeZoom;
+		if (newNativeZoom != oldNativeZoom) {
 			DPIUtil.setDeviceZoom (newNativeZoom);
-		}
 
-		notifyListeners(SWT.ZoomChanged, event);
+			float scalingFactor = 1f * DPIUtil.getZoomForAutoscaleProperty(newNativeZoom) / DPIUtil.getZoomForAutoscaleProperty(oldNativeZoom);
+			DPIZoomChangeRegistry.applyChange(this, newNativeZoom, scalingFactor);
 
-		if (getDisplay().isRescalingAtRuntime()) {
 			RECT rect = new RECT ();
 			COM.MoveMemory(rect, lParam, RECT.sizeof);
 			this.setBoundsInPixels(rect.left, rect.top, rect.right - rect.left, rect.bottom-rect.top);
+			return LRESULT.ZERO;
+		}
+	} else {
+		int newZoom = DPIUtil.getZoomForAutoscaleProperty (newNativeZoom);
+		int oldZoom = DPIUtil.getZoomForAutoscaleProperty (nativeZoom);
```

Comment:
This will change the behavior for the existing HiDPI support (without rescaling at runtime): in default mode `integer200`, a "zoom changed" event will not be generated for every native zoom change anymore, but only when the "corrected" one (100, 200 etc.) changes. In my opinion, this is a breaking API change (since consumers will receive the event in different cases than before), but in any way I propose to have such a change performed in a separate PR to be discussed there.

Reply:
Actually that is restoring the initial implementation. We adapted that in [77c1da](https://github.com/eclipse-platform/eclipse.platform.swt/commit/77c1da8956b7868e95eb45a3816ad732ceffda84#diff-f0c798cff8b637e98ce90c54208151a6a98e83f0c22c4e2427598957e108889c) some weeks ago

Reply:
Interesting, thank you for the pointer! Then it makes sense to take this chance and restore the existing behavior.